README.md
# Boleto Simples Ruby

[![Gem Version](http://img.shields.io/gem/v/boletosimples.svg)][gem]
[![Ruby](https://github.com/BoletoSimples/boletosimples-ruby/actions/workflows/ruby.yml/badge.svg?branch=master)](https://github.com/BoletoSimples/boletosimples-ruby/actions/workflows/ruby.yml)

[gem]: https://rubygems.org/gems/boletosimples

Biblioteca Ruby para acessar informações da [Kobana](https://www.kobana.com.br) através da [API](https://api.kobana.com.br).

## Instalação

Adicione a linha a baixo no seu Gemfile:

    gem 'boletosimples'

Execute:

    $ bundle install

Ou instale você mesmo:

    $ gem install boletosimples

## Configuração

Saiba mais sobre o [Token de API ](https://developers.kobana.com.br/reference/token-de-acesso)

```ruby
require 'boletosimples'

BoletoSimples.configure do |c|
  c.environment = :production # defaut :sandbox
  # production - https://app.kobana.com.br/conta/api/tokens
  # sandbox - https://app-sandbox.kobana.com.br/conta/api/tokens
  c.api_token = 'api-token'
  c.user_agent = 'email@minhaempresa.com.br' #Colocar um e-mail válido para contatos técnicos relacionado ao uso da API.
  # c.debug = true
  # c.custom_headers = { 'X-CUSTOM' => 'CONTENT' }
end
```

### Variáveis de ambiente

Você também pode configurar as variáveis de ambiente a seguir e não será necessário chamar `BoletoSimples.configure`

```bash
ENV['BOLETOSIMPLES_ENV']
ENV['BOLETOSIMPLES_API_TOKEN']
ENV['BOLETOSIMPLES_DEBUG']
```

### Configurando cache

É altamente recomendável utilizar o cache para evitar chegar no [limite de requisições](https://developers.kobana.com.br/reference/limite-de-requisições)

Para isso recomendamos a utilização da gem [Dalli](https://github.com/mperham/dalli)

Exemplo:

```ruby
  require 'dalli'

  BoletoSimples.configure do |c|
    c.cache = ActiveSupport::Cache.lookup_store(:mem_cache_store, ['localhost:11211'],
                                              namespace: 'boletosimples_client',
                                              compress: true)
  end
```

## Exemplos

### Boletos Bancários

```ruby
# Criar um boleto
@bank_billet = BoletoSimples::BankBillet.create(
  amount: 9.01,
  description: 'Despesas do contrato 0012',
  expire_at: '2014-01-01',
  customer_address: 'Rua quinhentos',
  customer_address_complement: 'Sala 4',
  customer_address_number: '111',
  customer_city_name: 'Rio de Janeiro',
  customer_cnpj_cpf: '012.345.678-90',
  customer_email: 'cliente@example.com',
  customer_neighborhood: 'Sao Francisco',
  customer_person_name: 'Joao da Silva',
  customer_person_type: 'individual',
  customer_phone_number: '2112123434',
  customer_state: 'RJ',
  customer_zipcode: '12312-123'
)

# Criar um novo boleto instanciando o objeto
@bank_billet = BoletoSimples::BankBillet.new(amount: 199.99, expire_at: '2020-01-01')
@bank_billet.description = 'Cobrança XPTO'
@bank_billet.save

# Mensagens de erro na criação do boleto
@bank_billet = BoletoSimples::BankBillet.create(amount: 199.99)
@bank_billet.response_errors
  # {:expire_at=>["não pode ficar em branco", "não é uma data válida"], :customer_person_name=>["não pode ficar em branco"], :customer_cnpj_cpf=>["não pode ficar em branco", "não é um CPNJ ou CPF válido"], :description=>["não pode ficar em branco"], :customer_zipcode=>["não pode ficar em branco"]

# Pegar informações de um boleto
@bank_billet = BoletoSimples::BankBillet.find(1) # onde 1 é o id do boleto.

 # Se o não for encontrado nenhum boleto com o id informado, uma exceção será levantada com a mensagem:
 # Couldn't find BankBillet with 'id'=1

# Listar os boletos
@bank_billets = BoletoSimples::BankBillet.all(page: 1, per_page: 50)
@bank_billets.each do |bank_billet|
  puts bank_billet.attributes
end

# Cancelar um boleto
@bank_billet = BoletoSimples::BankBillet.cancel(id: 1)

```

### Clientes

```ruby
# Criar um cliente
@customer = BoletoSimples::Customer.create(
  person_name: "Joao da Silva",
  cnpj_cpf: "012.345.678-90",
  email: "cliente@example.com",
  address: "Rua quinhentos",
  city_name: "Rio de Janeiro",
  state: "RJ",
  neighborhood: "bairro",
  zipcode: "12312-123",
  address_number: "111",
  address_complement: "Sala 4",
  phone_number: "2112123434"
)

# Mensagens de erro na criação do cliente
@customer = BoletoSimples::Customer.new(person_name: '')
@customer.response_errors
  # {:person_name=>["não pode ficar em branco"], :cnpj_cpf=>["não pode ficar em branco"], :zipcode=>["não pode ficar em branco"]}

# Listar os clientes
@customers = BoletoSimples::Customer.all(page: 1, per_page: 50)
@customers.each do |customer|
  puts customer.attributes
end

# Atualizar um cliente
@customer = BoletoSimples::Customer.find(1)
@customer.person_name = 'Novo nome'
@customer.save
```


CHANGELOG.md
# Changelog

## 2.1.0

- Implements CustomHeaders configuration

## 2.0.0

### Changes

- Update URLs based on new API: https://developers.kobana.com.br/reference/endpoints-1

## 1.0.4

### Changes

- Fix cache

## 1.0.3

### Changes

- Add debug option

## 1.0.0

### Breaking changes

- Remove Oauth code
- Require user_agent configuration
- Require api_token configuration
- All Errors now return with same root `errors`


example/oauth_example.rb
$LOAD_PATH << File.expand_path('../../lib', __FILE__)
require 'boletosimples'
require 'pp'

# If you need create users in Boleto Simples
# you can get a access token with app_id and app_secret
# def get_access_token
#   response = RestClient.post 'http://sandbox.boletosimples.com.br/api/v1/oauth2/token', {
#     grant_type: 'client_credentials',
#     client_id: 'app_id',
#     client_secret: 'app_secret'
#   }
#
#   access_app_token = JSON.parse(response)["access_token"]
# end

credentials = {
  token: 'access_app_token'
}

client_options = {
  user_agent: 'Meu e-Commerce (meuecommerce@example.com)'
}

client = BoletoSimples::OAuthClient.new('app_id',
                                        'app_secret',
                                        credentials,
                                        client_options)

# create user with oauth
# pp client.partner_create_user(
#   {
#     user: {
#       email: 'new_user@example.com'
#     }
#   }
# )

# Other operations need access tokens from user oatuh autorization process
# Use our omniauth-boletosimples (https://github.com/BoletoSimples/omniauth-boletosimples) to archive these tokens

credentials = {
  token: 'access token'
}

client_options = {
  user_agent: 'Meu e-Commerce (meuecommerce@example.com)'
}

client = BoletoSimples::OAuthClient.new('app_id',
                                        'app_secret',
                                        credentials,
                                        client_options)

# pp client.userinfo

# pp client.transactions

# pp client.customers

# pp client.customer 2
#
# pp client.create_customer(
#   {
#     person_name: "Joao da Silva",
#     cnpj_cpf: "012.345.678-90",
#     email: "cliente@bom.com",
#     address: "Rua quinhentos",
#     city_name: "Rio de Janeiro",
#     state: "RJ",
#     neighborhood: "bairro",
#     zipcode: "12312-123",
#     address_number: "111",
#     address_complement: "Sala 4",
#     phone_number: "2112123434"
#   }
# )

# pp client.bank_billets

# pp client.bank_billet 3

# Boleto normal
# pp client.create_bank_billet(
#   {
#     amount: 41.01,
#     customer_address: 'Rua quinhentos',
#     customer_address_complement: 'Sala 4',
#     customer_address_number: '111',
#     customer_city_name: 'Rio de Janeiro',
#     customer_cnpj_cpf: '012.345.678-90',
#     customer_email: 'cliente@bom.com',
#     customer_neighborhood: 'Sao Francisco',
#     customer_person_name: 'Joao da Silva',
#     customer_person_type: 'individual',
#     customer_phone_number: '2112123434',
#     customer_state: 'RJ',
#     customer_zipcode: '12312-123',
#     description: 'Despesas do contrato 0012',
#     expire_at: '2014-01-01',
#     notification_url: 'http://example.com.br/notify'
#   }
# )

# Boleto parcelado
# pp client.create_bank_billet(
#   {
#     amount: 41.01,
#     customer_address: 'Rua quinhentos',
#     customer_address_complement: 'Sala 4',
#     customer_address_number: '111',
#     customer_city_name: 'Rio de Janeiro',
#     customer_cnpj_cpf: '012.345.678-90',
#     customer_email: 'cliente@bom.com',
#     customer_neighborhood: 'Sao Francisco',
#     customer_person_name: 'Joao da Silva',
#     customer_person_type: 'individual',
#     customer_phone_number: '2112123434',
#     customer_state: 'RJ',
#     customer_zipcode: '12312-123',
#     description: 'Despesas do contrato 0012',
#     expire_at: '2014-01-01',
#     notification_url: 'http://example.com.br/notify',
#     parcels: 3
#   }
# )
